stock_picking_package_preparation_line/models/stock_picking_package_preparation.py

```diff
@@ -0,0 +1,215 @@
+# Copyright 2015 Francesco Apruzzese - Apulia Software srl
+# Copyright 2015-2018 Lorenzo Battistini - Agile Business Group
+# Copyright 2016 Alessio Gerace - Agile Business Group
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models, fields, api, _
+from datetime import datetime
+from odoo.exceptions import UserError
+from odoo.tools import float_compare
+
+
+class StockPickingPackagePreparation(models.Model):
+
+    _inherit = 'stock.picking.package.preparation'
+
+    FIELDS_STATES = {'done': [('readonly', True)],
+                     'in_pack': [('readonly', True)],
+                     'cancel': [('readonly', True)]}
+
+    picking_type_id = fields.Many2one(related='picking_ids.picking_type_id')
+    line_ids = fields.One2many(
+        'stock.picking.package.preparation.line',
+        'package_preparation_id',
+        string='Details',
+        copy=False,
+        states=FIELDS_STATES)
+
+    @api.multi
+    def _update_line_ids(self):
+        # Update package lines according to pickings.
+        # Lines with move_id will be updated
+        # Lines without move_id not be touched
+        line_model = self.env['stock.picking.package.preparation.line']
+        for pack in self:
+            for line in pack.line_ids:
+                if line.move_id:
+                    line.unlink()
+            package_preparation_lines = (
+                line_model._prepare_lines_from_pickings(
+                    [p.id for p in pack.picking_ids]))
+            if package_preparation_lines:
+                for line_vals in package_preparation_lines:
```

Comment:
@SimoRubi Just a remark, no change needed. Prefer building lines to add to package and do massive write with magic numbers (in v12, that will use multi create)

Reply:
thanks, noted for v12 migration :)